Feature<download method>

Liubomyr680 opened the pull request.

Liubomyr680 committed:
finish file download method

Liubomyr680 committed:
finish file download method

Liubomyr680 committed:
finish file download method

Liubomyr680 committed:
make GetMapping method to parse file

Liubomyr680 committed:
make PostMapping method to parse file

Liubomyr680 committed:
finish PostMapping method to parse file

Liubomyr680 committed:
finish PostMapping method to parse file

vshynkar commented on src/main/java/com/example/parser/utils/FileDownload.java:
```diff
@@ -0,0 +1,45 @@
+
+    public FileDownload(String fileUrl) throws IOException {
+        buffer = new byte[1024];
+        tmp = File.createTempFile("dozvoli-na-vikidi-onovlenii", ".tmp", (new File("/Users/kristina/Downloads/")));
```
Never do it. This code must work on any environment without code changes.

Liubomyr680 replied:
done

vshynkar commented on src/main/java/com/example/parser/service/PermitService.java:
```diff
@@ -0,0 +1,44 @@
+import java.util.List;
+
+@Service
+public class PermitService {
```
Service must contains endpoint processing logic. In another case it's not needed at all. Examle:
1) call DownloadFile code (singleton service or utility class) to download a file and save somewhere in temp files;
2) call Parser code (from appropriate singleton service) to parse downloaded data;
3) post-process parsed data (in case it's needed);
4) call repo code to save in database;
5) prepare endpoint response;

Liubomyr680 replied:
done

vshynkar commented on src/main/java/com/example/parser/utils/FileDownload.java:
```diff
@@ -0,0 +1,45 @@
+import java.net.HttpURLConnection;
+import java.net.URL;
+
+public class FileDownload {
```
It must be singleton service or utility class with static methods.

Liubomyr680 replied:
done

vshynkar commented on src/main/java/com/example/parser/utils/FileDownload.java:
```diff
@@ -0,0 +1,45 @@
+        return tmp;
+    }
+
+    public void download() throws IOException {
```
Handle IOException correctly. All opened resources and streams must be closed in case the exception was thrown.

Liubomyr680 replied:
done

vshynkar commented on src/main/java/com/example/parser/utils/Parser.java:
```diff
@@ -1,38 +1,47 @@
 import java.util.concurrent.TimeUnit;
 
+
 public class Parser {
```
This class must parse needed data. The code shouldn't think about downloading the data or preparation it for saving into database.

Liubomyr680 replied:
done

vshynkar commented on src/main/resources/script.ddl:
```diff
@@ -0,0 +1,13 @@
+create table json_data
```
Only nullable data fields should have nullable columns in database. Not nullable data must be saved in not null columns.

Liubomyr680 replied:
done

Liubomyr680 committed:
fix
1) Rename fields in ParsingResponse class
2) Rename ResponseFromParsing to ParsingResponse
3) add try() catch() to download() method
4) change file download path in FileDownload class

Liubomyr680 committed:
fix
Delete Service class
Divided the logic of the class Parser

Liubomyr680 committed:
fix Parser class

vshynkar commented on src/main/java/com/example/parser/repository/PermitRepository.java:
```diff
@@ -2,15 +2,12 @@
 import java.util.List;
 
 @Repository
 public interface PermitRepository extends JpaRepository<PermitForEmissionsOfPollutants, Long> {
```
Looks like this class has no code and can be deleted

Liubomyr680 replied:
done

vshynkar commented on src/main/java/com/example/parser/utils/FileDownload.java:
```diff
@@ -0,0 +1,51 @@
+        byte[] buffer = new byte[1024];
+
+        try {
+            tmpFile = File.createTempFile("dozvoli-na-vikidi-onovlenii", ".tmp");
```
Do not set a temporary file name manually. It will create a problem when few files will be downloaded simultaneously. Lets OS and Java set name.

Liubomyr680 replied:
done

vshynkar commented on src/main/java/com/example/parser/utils/FileDownload.java:
```diff
@@ -0,0 +1,51 @@
+                bout.write(buffer,0,read);
+
+        } catch (IOException e) {
+            log.error(e.getMessage());
```
Need to set some log message that can describe place of problem and set the exception object to see full stacktrace in logs. It could like this
`log.error("Can't download file with url={}", fileUrl, e)`

Liubomyr680 replied:
done

vshynkar commented on src/main/java/com/example/parser/utils/Parser.java:
```diff
@@ -55,22 +48,8 @@ public void parser(String filePath) throws IOException, ParseException {
-        log.info("2) Date the file was processed {}", date);
-        log.info("3) How many records were read from the file {}", fileCounter);
-        log.info("4) How many new records have been added to the database {}", dataBaseCounter);
+        return JsonList;
```
This variable name must start from lower case symbol.

Liubomyr680 replied:
done

vshynkar commented on src/main/java/com/example/parser/utils/StartProgram.java:
```diff
@@ -0,0 +1,65 @@
+        parsingResponse.setFileRecordsCounter(parser.getFileCounter());
+
+
+        List<PermitForEmissionsOfPollutants> dataFromDB = new ArrayList<>(permitRepository.findAll());
```
Do not read all table content just to compare with parsed data. There is a huge amount of data and it will require many RAM and CPU resources.

Liubomyr680 committed:
made last fixes

Liubomyr680 committed:
little fixes with ParserController, Parser, StartParsing class
